providers/unleash/src/main/java/dev/openfeature/contrib/providers/unleash/UnleashProvider.java

```diff
@@ -0,0 +1,173 @@
+        if (!ProviderState.READY.equals(state)) {
+            if (ProviderState.NOT_READY.equals(state)) {
+                throw new ProviderNotReadyError(PROVIDER_NOT_YET_INITIALIZED);
+            }
+            throw new GeneralError(UNKNOWN_ERROR);
+        }
+        UnleashContext context = ctx == null ? UnleashContext.builder().build() : ContextTransformer.transform(ctx);
+        boolean featureBooleanValue = unleash.isEnabled(key, context, defaultValue);
+        return ProviderEvaluation.<Boolean>builder()
+            .value(featureBooleanValue)
+            .build();
+    }
+
+    @Override
+    public ProviderEvaluation<String> getStringEvaluation(String key, String defaultValue, EvaluationContext ctx) {
+        ProviderEvaluation<Value> valueProviderEvaluation = getObjectEvaluation(key, new Value(defaultValue), ctx);
+        return ProviderEvaluation.<String>builder()
+            .value(valueProviderEvaluation.getValue().asString())
+            .variant(valueProviderEvaluation.getVariant())
+                .errorCode(valueProviderEvaluation.getErrorCode())
+                .reason(valueProviderEvaluation.getReason())
+                .flagMetadata(valueProviderEvaluation.getFlagMetadata())
+            .build();
+    }
+
+    @Override
+    public ProviderEvaluation<Integer> getIntegerEvaluation(String key, Integer defaultValue, EvaluationContext ctx) {
+        throw new TypeMismatchError(NOT_IMPLEMENTED);
+    }
+
+    @Override
+    public ProviderEvaluation<Double> getDoubleEvaluation(String key, Double defaultValue, EvaluationContext ctx) {
+        throw new TypeMismatchError(NOT_IMPLEMENTED);
+    }
+
+    @Override
+    public ProviderEvaluation<Value> getObjectEvaluation(String key, Value defaultValue, EvaluationContext ctx) {
+        if (!ProviderState.READY.equals(state)) {
+            if (ProviderState.NOT_READY.equals(state)) {
+                throw new ProviderNotReadyError(PROVIDER_NOT_YET_INITIALIZED);
+            }
+            throw new GeneralError(UNKNOWN_ERROR);
+        }
+        UnleashContext context = ctx == null ? UnleashContext.builder().build() : ContextTransformer.transform(ctx);
+        Variant evaluatedVariant = unleash.getVariant(key, context);
+        String variantName;
+        Value value;
+        if (DISABLED_VARIANT.equals(evaluatedVariant)) {
+            variantName = null;
+            value = defaultValue;
+        } else {
+            variantName = evaluatedVariant.getName();
+            value = evaluatedVariant.getPayload().map(p -> new Value(p.getValue())).orElse(null);
+        }
+        return ProviderEvaluation.<Value>builder()
+            .value(value)
+            .variant(variantName)
+            .build();
+    }
+
+    @Override
+    public void shutdown() {
+        super.shutdown();
+        log.info("shutdown");
+        if (unleash != null) {
+            unleash.shutdown();
+        }
+        state = ProviderState.NOT_READY;
```

Comment:
Does OpenFeature ever expect NOT_READY providers to reenter a READY state? If so, it might be worth allowing the GC to reclaim the unleash object, we definitely don't expect to be able to re initialize these, generally we make a new one

Reply:
It is not expected to use the provider for evaluations after it is shutdown.

Reply:
It's not specified if implementations should allow providers to re-enter NOT_READY/READY, so at this point it's up to the implementation.